src/slurm_plugin/instance_manager.py

```diff
@@ -619,3 +566,576 @@ def _create_request_for_nodes(table_name, node_names):
+            # instance to launch e.g.
+            # {
+            #   queue_1: {cr_1: [nodes_1, nodes_2, nodes_3], cr_2: [nodes_4]},
+            #   queue_2: {cr_3: [nodes_5]}
+            # }
+
+            instances_launched = self._launch_instances_for_job(
+                job=job,
+                nodes_to_launch=nodes_to_launch,
+                launch_batch_size=launch_batch_size,
+                all_or_nothing_batch=all_or_nothing_batch,
+            )
+            # instances_launched e.g.
+            # {
+            #   queue_1: {cr_1: list[EC2Instance], cr_2: list[EC2Instance],
+            #   queue_2: {cr_3: list[EC2Instance]}
+            # }
+
+            if sum(
+                len(instance_list)
+                for compute_resources in instances_launched.values()
+                for instance_list in compute_resources.values()
+            ) == len(job.nodes_resume):
+                # All requested capacity for the Job has been launched
+                # Assign launched EC2 instances to the requested Slurm nodes
+                try:
+                    logger.info(
+                        "Successful launched all instances for nodes %s",
+                        print_with_count(job.nodes_resume),
+                    )
+                    self._assign_instances_to_nodes(
+                        update_node_address=update_node_address,
+                        nodes_to_launch=nodes_to_launch,
+                        instances_launched=instances_launched,
+                        update_node_batch_size=update_node_batch_size,
+                    )
+                except (HostnameDnsStoreError, HostnameTableStoreError):
+                    # Failed to assign EC2 instances to nodes
+                    # EC2 Instances already assigned, are going to be terminated by
+                    # setting the nodes into DOWN.
+                    # EC2 instances not yet assigned, are going to fail during bootstrap,
+                    # because no entry in the DynamoDB table would be found
+                    self._update_failed_nodes(set(job.nodes_resume))
+            elif instances_launched:
+                # Try to reuse partial capacity of already launched EC2 instances
+                logger.info(
+                    "Launched instances that can be reused %s",
+                    print_with_count(
+                        [
+                            (queue, compute_resource, instance)
+                            for queue, compute_resources in instances_launched.items()
+                            for compute_resource, instances in compute_resources.items()
+                            for instance in instances
+                        ]
+                    ),
+                )
+                self.unused_launched_instances |= instances_launched
+                self._update_failed_nodes(set(job.nodes_resume), "LimitedInstanceCapacity")
```

Comment:
Should we consider centralising the error code names (`LimitedInstanceCapacity` and `InsufficientInstanceCapacity`)?

Reply:
I would consider it in another PR